### PR TITLE
MessagesDataSource & MessagesLayoutDelegate methods for MessageType.custom

### PR DESCRIPTION
**Pull request by RealBonus**
What does this implement/fix? Explain your changes.
---------------------------------------------------
Simpler solution for working with custom cells: Delegate methods. No needs to override or implement custom FlowLayout's.

Any other comments?
-------------------
👻

Where has this been tested?
---------------------------
**iOS Version:**  11, 12
Not a major logic refactoring, should work anywhere.

### CODE REVIEW

File: Sources/Protocols/MessagesLayoutDelegate.swift [hunk 1]
```diff
@@ -80,6 +80,17 @@ public protocol MessagesLayoutDelegate: AnyObject {
...
+    /// Cell size calculator for messages with MessageType.custom.
+    ///
+    /// - Parameters:
+    ///   - message: The custom message
+    ///   - indexPath: The `IndexPath` of the cell.
+    ///   - messagesCollectionView: The `MessagesCollectionView` in which this cell will be displayed.
+    ///
+    /// - Note:
+    ///   The default implementation will throw fatalError(). You must override this method if you are using messages with MessageType.custom.
+    func cellSizeCalculator(for message: MessageType, at indexPath: IndexPath, in messagesCollectionView: MessagesCollectionView) -> CellSizeCalculator
```
Comment (SD10) [inline Sources/Protocols/MessagesLayoutDelegate.swift]: Hey @RealBonus, this is an interesting PR. Thanks for taking the time to submit it 👍 I'm a little concerned that this method is called `cellSizeCalculator` but is only used for the `.custom` case of `MessageKind`. Maybe `customCellSizeCalculator` is a better name? Aside from that, I'll let some other team members review this

Comment (RealBonus) [reply Sources/Protocols/MessagesLayoutDelegate.swift]: Good point, agreed

Comment (nathantannar4) [reply Sources/Protocols/MessagesLayoutDelegate.swift]: I think this does a good job making it easier for people to get started with using custom cells with, as you said, not having to subclass the layout. And if people still want to subclass the layout they can, this doesn't raise any breaking changes.

**Review by nathantannar4 — APPROVED** [review]

File: Sources/Layout/MessagesCollectionViewFlowLayout.swift [hunk 2]
```diff
@@ -151,7 +151,7 @@ open class MessagesCollectionViewFlowLayout: UICollectionViewFlowLayout {
         case .location:
             return locationMessageSizeCalculator
         case .custom:
-            fatalError("Must return a CellSizeCalculator for MessageKind.custom(Any?)")
+            return messagesLayoutDelegate.customCellSizeCalculator(for: message, at: indexPath, in: messagesCollectionView)
```
Comment (SD10) [inline Sources/Layout/MessagesCollectionViewFlowLayout.swift:156]: I think we need to add a comment to tell the user that if they override
`func cellSizeCalculatorForItem(at indexPath: IndexPath) -> CellSizeCalculator`
they will then be responsible for calling the `customCellSizeCalculator` method if necessary

File: Sources/Controllers/MessagesViewController.swift [hunk 3]
```diff
@@ -206,7 +206,7 @@ UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {
             cell.configure(with: message, at: indexPath, and: messagesCollectionView)
             return cell
         case .custom:
-            fatalError(MessageKitError.customDataUnresolvedCell)
+            return messagesDataSource.customCell(for: message, at: indexPath, in: messagesCollectionView)
```
Comment (SD10) [inline Sources/Controllers/MessagesViewController.swift:211]: Same here, I'm worried someone overrides `cellForItem(at:)` and then doesn't handle calling `customCell`. Then they are using the delegate and expecting it to be called

**Comment by SD10** [comment]
@RealBonus We also need to add a CHANGELOG entry for these changes

**Comment by nathantannar4** [comment]
@RealBonus Can you please add a CHANGELOG entry and add the comments Steven requested so we can get this in the next release!

**Comment by RealBonus** [comment]
Sorry for delay guys, out of town out of business.
CHANGELOG and comments: Done!

**Comment by nathantannar4** [comment]
🔥 💯

**Comment by SD10** [comment]
Thank you for contributing to MessageKit! I've invited you to join the MessageKit GitHub organization - no pressure to accept! If you'd like more information on what that means, check out our [contributing guidelines](https://github.com/MessageKit/MessageKit/blob/master/CONTRIBUTING.md) and join the [MessageKit Slack channel](https://join.slack.com/t/messagekit/shared_invite/enQtMjI4NzIzNzMyMzU0LTEyNmYzNzBlNjNiYjAwYjY1ZjhkMmVjNDE4OGNlMDY1OTRmNzVmMzI4MTg3OWQzZTM2NGU4NTU3ZDE5NzkwNzQ). Feel free to reach out if you have any questions! :smiley:

**Pull request merged**